src/rail/estimation/algos/random_forest.py

```diff
@@ -0,0 +1,166 @@
+            # We also use the colours as training data, even the redundant ones
+            for b2 in self.config.class_bands[:]:
+                b2_cat=self.config.bands[b2]
+                if b1 < b2:
+                    features.append(f"{b1}-{b2}")
+                    training_data.append(training_data_table[b1_cat] - training_data_table[b2_cat])
+        training_data = np.array(training_data).T
+
+        print("Training data for bin classifier has shape ", training_data.shape)
+
+        # Now put the training data into redshift bins
+        # We use -99 to indicate that we are outside the desired ranges
+        z = training_data_table[self.config.redshift_col]
+        training_bin = np.repeat(self.config.no_assign, len(z))
+        print("Using these bin edges:", self.config.bin_edges)
+        for i, zmin in enumerate(self.config.bin_edges[:-1]):
+            zmax = self.config.bin_edges[i + 1]
+            training_bin[(z > zmin) & (z < zmax)] = i
+            ntrain_bin = ((z > zmin) & (z < zmax)).sum()
+            print(f"Training set: {ntrain_bin} objects in tomographic bin {i}")
+
+        # Can be replaced with any classifier
+        skl_classifier = skl_RandomForestClassifier(
+            max_depth=10,
+            max_features=None,
+            n_estimators=20,
+            random_state=self.config.random_seed,
+        )
+        skl_classifier.fit(training_data, training_bin)
+
+        #return classifier, features
+        self.model = randomForestmodel(skl_classifier, features)
+        self.add_data('model', self.model)
+        
+
+class randomForestClassifier(CatClassifier):
+    """Classifier that assigns tomographic 
+    bins based on random forest method"""
+    
+    name = 'randomForestClassifier'
+    config_options = CatClassifier.config_options.copy()
+    config_options.update(
+        id_name=Param(str, "", msg="Column name for the object ID in the input data, if empty the row index is used as the ID."),
+        class_bands=Param(tuple, ["r","i","z"], msg="Which bands to use for classification"),
+        bands=Param(dict, {"r":"mag_r_lsst", "i":"mag_i_lsst", "z":"mag_z_lsst"}, msg="column names for the the bands"),)
+    outputs = [('output', Hdf5Handle)]
+    
+    def __init__(self, args, comm=None):
+        CatClassifier.__init__(self, args, comm=comm)
+            
+            
+    def open_model(self, **kwargs):
+        CatClassifier.open_model(self, **kwargs)
+        if self.model is None:  # pragma: no cover
+            return
+        self.skl_classifier = self.model.skl_classifier
+        self.features = self.model.features
+
+    
+    def run(self):
+        """Apply the classifier to the measured magnitudes"""
+        
+        if self.config.hdf5_groupname:
+            test_data = self.get_data('input')[self.config.hdf5_groupname]
+        else:  # pragma: no cover
+            test_data = self.get_data('input')
+        
+        data = []
+        for f in self.features:
```

Comment:
This mostly repeats lines 59-69 so I think would be better off as a standalone `calculate_colors` function that both classes will call. Then in the future we can consider moving that function to a rail_base utility module since it's repeated in other algos as well.

Reply:
The two blocks seem to have different input/output, so will be kept as they are.